src/errors/typeErrors.ts

```diff
@@ -1,11 +1,111 @@
 import * as es from 'estree'
-import { ErrorSeverity, ErrorType, SourceError, Type, TypeAnnotatedNode } from '../types'
+import { ErrorSeverity, ErrorType, SourceError, SArray, Type, TypeAnnotatedNode } from '../types'
 import { simplify, stripIndent } from '../utils/formatters'
 import { typeToString } from '../utils/stringify'
 import { generate } from 'astring'
 
 // tslint:disable:max-classes-per-file
 
+export class InvalidArrayIndexType implements SourceError {
+  public type = ErrorType.TYPE
+  public severity = ErrorSeverity.WARNING
+
+  constructor(public node: TypeAnnotatedNode<es.Node>, public receivedType: Type) {}
+
+  get location() {
+    return this.node.loc!
+  }
+
+  public explain() {
+    return `Expected array index as number, got ${typeToString(this.receivedType)} instead`
+  }
+
+  public elaborate() {
+    return this.explain()
+  }
+}
+
+export class ArrayAssignmentError implements SourceError {
+  public type = ErrorType.TYPE
+  public severity = ErrorSeverity.WARNING
+
+  constructor(
+    public node: TypeAnnotatedNode<es.Node>,
+    public arrayType: SArray,
+    public receivedType: Type
+  ) {}
+
+  get location() {
+    return this.node.loc!
+  }
+
+  public explain() {
+    return stripIndent`Array expected type: ${typeToString(this.arrayType)}
+    but got: ${typeToString(this.receivedType)}`
+  }
+
+  public elaborate() {
+    return this.explain()
+  }
+}
+
+export class ReassignConstError implements SourceError {
+  public type = ErrorType.TYPE
+  public severity = ErrorSeverity.WARNING
+
+  constructor(public node: TypeAnnotatedNode<es.AssignmentExpression>) {}
+
+  get location() {
+    return this.node.loc!
+  }
+
+  public explain() {
+    const [varName] = formatAssignment(this.node)
+    return `Reassignment of constant ${varName}`
+  }
+
+  public elaborate() {
+    return this.explain()
+  }
+}
+
+export class DifferentAssignmentError implements SourceError {
+  public type = ErrorType.TYPE
+  public severity = ErrorSeverity.WARNING
+
+  constructor(
+    public node: TypeAnnotatedNode<es.AssignmentExpression>,
+    public expectedType: Type,
+    public receivedType: Type
+  ) {}
+
+  get location() {
+    return this.node.loc!
+  }
+
+  public explain() {
+    const [varName, assignmentStr] = formatAssignment(this.node)
+    return stripIndent`
+    Expected reassignment of ${varName}:
```

Comment:
How about:
Expected assignment of ${varName}:
...
to get a value of type:
...
but got a value of type:
...

Reply:
Yes, that is easier to read. Thanks for the suggestion!

Reply:
I suggested the use of the term "assignment" and not "reassignment". Any assignment is a "reassignment", so the "re" doesn't add anything.